mysql-test/t/blackhole.test

```diff
@@ -41,3 +41,38 @@ SELECT MATCH (a) AGAINST ('abc') AS score FROM t;
 # But this index scan failed. Should succeed now.
 SELECT 1 FROM t WHERE MATCH (a) AGAINST ('abc');
 DROP TABLE t;
+
```

Comment:
This only tests non-transactional blackhole: please set the plugin var first.

Consider converting this snippet to an include file, and testing both false and true for the sysvar.

Reply:
It looks like even before implementing transactions, BEGIN TRANSACTION and COMMIT were allowed and it would just no-op. So with blackhole_transactions on or off, they both yielded the same test results. 

Do you know of a way to actually ensure commit does something?

I tried to use SHOW BINLOG EVENTS via --source include/show_binlog_events.inc but whether blackhole_transactions is on or off, it produces the same binlog events. What's curious is that, even with:

START TRANSACTION;
... 
ROLLBACK;

they produce the same binlog events. So it seems like rollback works even without transaction support?

Reply:
> Do you know of a way to actually ensure commit does something?

Check the binlog comment below. If that does not help, maybe the `Handler_commit` status variable will be bumped more? If that doesn't help, I would still consider testing both variable settings even without user-visible changes between them: it would test that the server does not assert etc.

> I tried to use SHOW BINLOG EVENTS via --source include/show_binlog_events.inc but whether blackhole_transactions is on or off, it produces the same binlog events. What's curious is that, even with:
> 
> START TRANSACTION; ... ROLLBACK;
> 
> they produce the same binlog events. So it seems like rollback works even without transaction support?

Could you please re-check whether that's the case with transactional tables? AFAIK, binlog has different behavior for transactions that mix-in non-transactional SE changes:
1) If all SEs in a transaction are transactional, then the transaction gets binlogged on COMMIT, and on ROLLBACK nothing gets binlogged
2) If there is at least one non-transcational SE in a transaction, then the transaction gets binlogged (immediately after each statement? not sure) always, and on ROLLBACK the ROLLBACK statement itself gets binlogged. This way replicas can try to reproduce what happened on the primary.

The behavior you described matches 2) but with transactional Blackhole 1) must also happen.

Reply:
What I find interesting is that without my changes to the blackhole engine, start transaction and rollback seems to work, as in, nothing gets binlogged due to the rollback. Looking at the blackhole documentation, this statement seems to hold true in my tests:

"The BLACKHOLE engine is transaction-aware, in the sense that committed transactions are written to the binary log and rolled-back transactions are not."
https://dev.mysql.com/doc/refman/8.0/en/blackhole-storage-engine.html

I tested another non-transactional engine such as Memory and found that the rollback statement does nothing. So, there's something in blackhole implementation that is making it "transaction aware", even before my changes.

However, the most important thing we add with my changes is that blackhole transactions actually block on commit, which it didn't before.